### PR TITLE
[Impeller] Patch the compiler to account for subpass inputs and PSO metadata.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Towards https://github.com/flutter/flutter/issues/128911

Drive by fixes https://github.com/flutter/flutter/issues/123795

### CODE REVIEW

File: impeller/aiks/aiks_unittests.cc [hunk 1]
```diff
@@ -1636,7 +1636,7 @@ TEST_P(AiksTest, ColorWheel) {
   auto callback = [&](AiksContext& renderer) -> std::optional<Picture> {
     // UI state.
     static bool cache_the_wheel = true;
-    static int current_blend_index = 3;
+    static int current_blend_index = 14;
```
Comment (bdero) [inline impeller/aiks/aiks_unittests.cc:1639]: Nit: Keep default to SourceOver? Although advanced blends are more contentious, so maybe someone flipping through is more likely to notice a problem with Screen as the default.

File: impeller/compiler/reflector.cc [hunk 2]
```diff
@@ -261,6 +262,19 @@ std::optional<nlohmann::json> Reflector::GenerateTemplateArguments() const {
     }
   }
 
+  {
+    if (auto inputs = ReflectResources(shader_resources.subpass_inputs);
```
Comment (bdero) [inline impeller/compiler/reflector.cc:266]: Oh, convenient!

**Review by bdero — APPROVED** [review]
LGTM

Would a reflection test for this in `compiler_unittests.cc` make sense?

File: impeller/aiks/aiks_unittests.cc (hunk 1 above)
Comment (chinmaygarde) [reply impeller/aiks/aiks_unittests.cc:1639]: Yeah, just thought defaulting to the advanced ones was more likely to throw up issues.

**Comment by chinmaygarde** [comment]
> Would a reflection test for this in compiler_unittests.cc make sense?

My hope is that the pipeline creation itself would be the ultimate fix. Another patch incoming that wires up the advanced blend libraries.

File: impeller/compiler/compiler.cc [hunk 3]
```diff
@@ -115,15 +115,13 @@ static CompilerBackend CreateMSLCompiler(
...
-  // TODO(dnfield): It seems like what we'd want is a CompilerGLSL with
-  // vulkan_semantics set to true, but that has regressed some things on GLES
-  // somehow. In the mean time, go back to using CompilerMSL, but set the Metal
-  // Language version to something really high so that we don't get weird
-  // complaints about using Metal features while trying to build Vulkan shaders.
-  // https://github.com/flutter/flutter/issues/123795
-  return CreateMSLCompiler(
-      ir, source_options,
-      spirv_cross::CompilerMSL::Options::make_msl_version(3, 0, 0));
+  auto vk_compiler = std::make_shared<spirv_cross::CompilerGLSL>(ir);
```
Comment (jonahwilliams) [inline impeller/compiler/compiler.cc:118]: Do we know how this change had previously broken GLES and how this changed fixed it?

Comment (chinmaygarde) [reply impeller/compiler/compiler.cc:118]: The original issue was light on details but there is now a separate `CreateGLSLCompiler` which is probably what's used. So this change shouldn't affect GLSL at all. I believe earlier there used to be a common compiler instance creation and I can believe was hard to configure just right for one backend without causing unintended consequences elsewhere. But again, the issue was light on details so its hard to say. I'll keep an eye out for repercussions.

**Pull request merged**